src/abstract_domains.rs

```diff
@@ -291,20 +326,46 @@ impl AbstractDomain for ExpressionDomain {
     /// Returns an expression that is "self == other".
     fn equals(&self, other: &Self) -> Self {
         match (self, other) {
+            // If self and other are constant values, just compare them with ==.
             (
                 ExpressionDomain::CompileTimeConstant(cv1),
                 ExpressionDomain::CompileTimeConstant(cv2),
             ) => {
-                if cv1 == cv2 {
+                return if cv1 == cv2 {
                     ExpressionDomain::CompileTimeConstant(ConstantValue::True)
                 } else {
                     ExpressionDomain::CompileTimeConstant(ConstantValue::False)
+                };
+            }
+            // If self and other are the same location in memory, return true unless the value might be NaN.
+            (
+                ExpressionDomain::Variable {
+                    path: p1,
+                    var_type: t1,
+                },
+                ExpressionDomain::Variable {
+                    path: p2,
+                    var_type: t2,
+                },
+            ) => {
+                if (**p1) == **p2 {
+                    match (t1, t2) {
+                        (ExpressionType::F32, _)
```

Comment:
These functions are on the cusp of getting too long and intended for me. Consider having some functions like is_floating_point_expression or whatever.

Reply:
Ooo, they are going to get a lot longer before long. Let me think it over a little later on.